src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs

```diff
@@ -86,55 +118,158 @@ public ValidationRuleSet(ValidationRuleSet ruleSet)
         /// </summary>
         /// <param name="rules">Rules to be contained in this ruleset.</param>
-        public ValidationRuleSet(IEnumerable<ValidationRule> rules)
+        public ValidationRuleSet(IDictionary<string, IList<ValidationRule>> rules)
         {
             if (rules == null)
             {
                 return;
             }
 
-            foreach (ValidationRule rule in rules)
+            foreach (var rule in rules)
             {
-                Add(rule);
+                Add(rule.Key, rule.Value);
             }
         }
 
         /// <summary>
-        /// Gets the rules in this rule set.
+        /// Add the new rule into the rule set.
         /// </summary>
-        public IEnumerable<ValidationRule> Rules
+        /// <param name="key">The key for the rule.</param>
+        /// <param name="rules">The list of rules.</param>
+        public void Add(string key, IList<ValidationRule> rules)
         {
-            get
+            foreach (var rule in rules)
             {
-                return _rules.Values.SelectMany(v => v);
+                Add(key, rule);
             }
         }
 
         /// <summary>
-        /// Add the new rule into the rule set.
+        /// Add a new rule into the rule set.
         /// </summary>
+        /// <param name="key">The key for the rule.</param>
         /// <param name="rule">The rule.</param>
-        public void Add(ValidationRule rule)
+        /// <exception cref="OpenApiException">Exception thrown when rule already exists.</exception>
+        public void Add(string key, ValidationRule rule)
         {
-            if (!_rules.ContainsKey(rule.ElementType))
+            if (!_rules.ContainsKey(key))
             {
-                _rules[rule.ElementType] = new List<ValidationRule>();
+                _rules[key] = new List<ValidationRule>();
             }
 
-            if (_rules[rule.ElementType].Contains(rule))
+            if (_rules[key].Contains(rule))
             {
                 throw new OpenApiException(SRResource.Validation_RuleAddTwice);
             }
 
-            _rules[rule.ElementType].Add(rule);
+            _rules[key].Add(rule);
+        }
+
+        /// <summary>
+        /// Updates an existing rule with a new one.
+        /// </summary>
+        /// <param name="key">The key of the existing rule.</param>
+        /// <param name="newRule">The new rule.</param>
+        /// <param name="oldRule">The old rule.</param>
+        /// <returns>true, if the update was successful; otherwise false.</returns>
+        public bool Update(string key, ValidationRule newRule, ValidationRule oldRule)
+        {
+            if (!_rules.ContainsKey(key))
+            {
+                return false;
+            }
+            else
+            {
+                _rules[key].Remove(oldRule);
+                _rules[key].Add(newRule);
+                return true;
+            }
```

Comment:
```suggestion
            if (_rules.TryGetValue(key, out var currentRules))
            {
                currentRules.Add(newRule);
                return currentRules.Remove(oldRule);
            }
            return false;
```

Note: this implementation doesn't conserve order, but the previous one didn't either, I'm not sure whether it's a requirement?

Reply:
I don't believe the order should matter really.